docs/developer-tools/ThrowHelper.md

```diff
@@ -0,0 +1,165 @@
+- [SynchronizationLockException](https://docs.microsoft.com/dotnet/api/system.threading.SynchronizationLockException)
+- [TimeoutException](https://docs.microsoft.com/dotnet/api/system.TimeoutException)
+- [UnauthorizedAccessException](https://docs.microsoft.com/dotnet/api/system.UnauthorizedAccessException)
+- [Win32Exception](https://docs.microsoft.com/dotnet/api/system.componentmodel.Win32Exception)
+
+You can see the available constructors for each of these exception types from the linked docs - the respective `ThrowHelper` APIs will simply offer a series of overloads corresponding to the existing constructors for that specified type. Each API is simply named "Throw" + the exception type.
+
+## Technical Details
+
+The following section describes the impact of switching to `ThrowHelper` APIs over the standard `throw new Expression` statements with respect to the codegen. As mentioned above, knowing these details is not necessary in order to use these APIs, but for developers familiar with how the runtime works or with how the JIT processes code during execution, this will provide a clearer explanation of why using the `ThrowHelper` pattern results in more compact and faster code. In order to do so, we will need to have a peek at the code that the JIT compiler is actually producing in these situations. We will use the code generated on .NET Core 3.1 x64 as a reference, but the same concept applies to other runtimes and architectures as well.
+
+Let's consider this simple type:
+
+```csharp
+public struct ExceptionTest
+{
+    private int number;
+    
+    public int GetValueIfNotZeroOrThrow()
+    {
+        if (number == 0)
+        {
+            throw new InvalidOperationException("The value must be != 0");
+        }
+        
+        return number;
+    }
+}
+```
+
+This type is only meant to showcase the use of `ThrowHelper` APIs. Here we're simply checking the value of the `number` field, and throwing if it's 0. Let's check the generated asm code:
+
+```x86asm
+ExceptionTest.GetValueIfNotZeroOrThrow()
+    mov eax, [rcx]              ; read number
+    test eax, eax               ; if (number != 0)
+    je short L0011              ; jump to faulting path (if == 0)
+    ret                         ; return (number is already in eax)
+    mov rcx, 0x7ff95cbaca80     ; exception setup...
+    call 0x00007ff9bc6178f0
+    mov rsi, rax
+    mov ecx, 1
+    mov rdx, 0x7ff96590c080
+    call 0x00007ff9bc7403e0
+    mov rdx, rax
+    mov rcx, rsi
+    call System.InvalidOperationException..ctor(System.String)
+    mov rcx, rsi
+    call 0x00007ff9bc5db3a0
+    int3 ; finally throw the exception
+```
+
+We can see the code contains a lot of lines just dedicated to creating the exception being thrown. This can cause a large increase in the code size whenever we have many of these checks, because it will reduce the efficiency of the instruction cache, etc. In general, we'd like the size of our methods to be as small as possible.
```

Comment:
```suggestion
We can see the code contains a lot of lines just dedicated to creating the exception being thrown. This can cause a large increase in the code size whenever we have many of these checks because it will reduce the efficiency of the instruction cache, etc. In general, we'd like the size of our methods to be as small as possible.
```